bignumber.js

```diff
@@ -2369,14 +2369,46 @@
          *
          * 'pow() exponent not an integer: {n}'
          * 'pow() exponent out of range: {n}'
+         *
+         * m {number} Integer|BigNumber
          */
-        P.toPower = P.pow = function (n) {
+        P.toPower = P.pow = function (n, m) {
             var k, y,
                 i = mathfloor( n < 0 ? -n : +n ),
-                x = this;
+                x = this,
+                nIsValidInt = isValidInt( n, -MAX_SAFE_INTEGER, MAX_SAFE_INTEGER, 23, 'exponent' ),
+                hasMod = typeof m !== 'undefined';
+
+            //Implement fast modular exponentiation via
+            //RTL binary method when n >= 0 && m > 0
+            //Returns a BigNumber which equals (x ^ n) % m
+            function powMod(x, n, m, k) {
+                var y = new BigNumber(ONE);
+
+                //For this algorithm to work with n < 0,
+                //must calculate the modular inverse of x (mod m).
+                if ( n < 0 ) {
+                    raise(23, 'modular exponentiation with n < 0 not supported.', n);
+                }
+
+                //Any int modulus 1 is 0
+                if ( m === 1 ) return 0;
+
+                x = x.mod( m );
+                while ( n > 0 ) {
+                    if ( n % 2 === 1 ) {
+                        y = y.times( x ).mod( m );
+                        if ( k && y.c.length > k ) y.c.length = k;
+                    }
+                    n = n >> 1;
```

Comment:
Remember `n` can be up a value up to `9007199254740992`, and this bitwise halving will fail for numbers over 2^31, which is why I used `mathfloor( i / 2 )`.